include/alpaka/core/CallbackThread.hpp

```diff
@@ -34,46 +37,65 @@ namespace alpaka::core
                 m_thread.join();
             }
         }
-        auto submit(Task&& newTask) -> std::future<void>
+
+        // Note: due to different std lib implementations of packaged_task, the lifetime of the passed function either
+        // ends when the packaged_task is destroyed (while the returned future is still alive) or when both are
+        // destroyed.
+        template<typename NullaryFunction>
+        auto submit(NullaryFunction&& nf) -> std::future<void>
         {
-            auto f = newTask.get_future();
+            static_assert(
+                std::is_void_v<std::invoke_result_t<NullaryFunction>>,
+                "Submitted function must not have any arguments and return void.");
+            auto task = Task{std::forward<NullaryFunction>(nf)};
+            auto f = task.get_future();
             {
                 std::unique_lock<std::mutex> lock{m_mutex};
-                m_tasks.emplace(std::move(newTask));
+                ++m_tasksInProgress;
+                m_tasks.emplace(std::move(task));
                 if(!m_thread.joinable())
                     startWorkerThread();
             }
             m_cond.notify_one();
             return f;
         }
 
+        [[nodiscard]] auto empty() const
+        {
+            return m_tasksInProgress == 0;
```

Comment:
Is is not better to use the mutex here and check the queue if they is empty or not. Doing a second accounting even with an atomic variable will most likely introduce a moment where the queue and the counter are not synced.
The counter is only introducing another complexity.

I suggest


```C++
std::unique_lock<std::mutex> lock{m_mutex};
m_tasks.empty();
```

Reply:
Yeah, that was my first attempt as well! However, consider that when the background thread dequeues the last task and starts running it, the `m_tasks` is already empty. So if you query the alpaka queues on whether they are empty inside a host func, they say true, but obviously you are still running a task :)

This case is also covered by the unit tests btw.